### PR TITLE
Move croquemort features from udata to udata-croquemort

### PR DESCRIPTION
**Pull request by abulte**
Associated with https://github.com/opendatateam/udata/pull/1099.

Proof of concept, to be iterated and discussed. The coupling is ugly and some features are missing (i18n). The current JS won't work on some browsers.

### CODE REVIEW

File: udata_croquemort/__init__.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from flask import render_template, Blueprint
+
+from udata.frontend import footer_snippet
+
+blueprint = Blueprint('piwik', __name__, template_folder='templates')
```
Comment (davidbgk) [inline udata_croquemort/__init__.py]: croquemort?

File: udata_croquemort/templates/croquemort.html [hunk 2]
```diff
@@ -0,0 +1,44 @@
+<script type="text/javascript">
+    const check_urls = {{ 'true' if config.CROQUEMORT else 'false' }};
```
Comment (davidbgk) [inline udata_croquemort/templates/croquemort.html]: I'm not sure you can use that much ES6 without turning it to ES5 at some point.

File: udata_croquemort/__init__.py (hunk 1 above)
Comment (abulte) [reply udata_croquemort/__init__.py]: 😨

File: udata_croquemort/templates/croquemort.html (hunk 2 above)
Comment (abulte) [reply udata_croquemort/templates/croquemort.html]: Yep, just a prototype ;-) Fetch is supported everywhere either.

**Comment by abulte** [comment]
New version (https://github.com/opendatateam/udata-croquemort/pull/2/commits/81df5dd60c17ae612a1e36cf44aa3197b3a3f548) is quite different from first draft:
- frontend related stuff stay in `udata`
- udata-croquemort exposes a `linkchecker` entrypoint which is used by `udata` to trigger a synchronous check — in the future the link checker will be configurable on a resource basis

Tests are still TODO, I'm having a strange problem running them, cf Slack.

File: CHANGELOG.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# Changelog
+
+## Current (in progress)
```
Comment (davidbgk) [inline CHANGELOG.md]: Maybe a `0.0.1` here?

File: README.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
+# uData-croquemort
+
+This plugin provide integration between uData and Croquemort link checker.
```
Comment (davidbgk) [inline README.md]: provides*

File: udata_croquemort/checker.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        log.debug('Checking resource w/ URL %s' % resource.url)
+        # XXX: is this a sane and reliable way to get the dataset?
+        dataset = resource._instance
+        try:
+            check_response = check_url(resource.url, group=dataset.slug)
+            return self._format_response(check_response)
+        except UnreachableLinkChecker as e:
+            log.error('Unreachable croquemort for url {}: {}'.format(
+                      resource.url, e))
+            pass
```
Comment (davidbgk) [inline udata_croquemort/checker.py]: `return None` looks more explicit.

File: udata_croquemort/checker.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+    The main interface is the `check` method.
+    """
+
+    def _format_response(self, response):
+        status = response.get('final-status-code')
+        try:
+            status = int(status)
+            return {
+                'check:url': response.get('checked-url'),
```
Comment (davidbgk) [inline udata_croquemort/checker.py]: Nice namespacing, do we expect anything that is not prefixed by `check:`?

File: udata_croquemort/utils.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+class UnreachableLinkChecker(Exception):
+    pass
+
+
+def is_pending(response):
+    if not response:
+        return True
+    if response.status_code == httplib.NOT_FOUND:
+        return True
+    try:
```
Comment (davidbgk) [inline udata_croquemort/utils.py]: Why `try/except` vs. regular `if` check here?

File: CHANGELOG.md (hunk 3 above)
Comment (abulte) [reply CHANGELOG.md]: I think `bumpr` will do this for me when we release, right @noirbizarre?

File: README.md (hunk 4 above)
Comment (abulte) [reply README.md]: 👍

File: udata_croquemort/checker.py (hunk 6 above)
Comment (abulte) [reply udata_croquemort/checker.py]: Nope, all `check:`

File: udata_croquemort/checker.py (hunk 5 above)
Comment (abulte) [reply udata_croquemort/checker.py]: 👍

File: udata_croquemort/utils.py (hunk 7 above)
Comment (abulte) [reply udata_croquemort/utils.py]: Not my code ;-) But yeah, I don't see why not.

Comment (davidbgk) [reply udata_croquemort/utils.py]: 🎉 Achievement unlocked: I progressed since then 😉

Comment (abulte) [reply udata_croquemort/utils.py]: Ah ah it was you? 😛

**Comment by abulte** [comment]
@noirbizarre @davidbgk if it's OK with you, I'm gonna merge this to `master` and work on a new PR for tooling and deployment stuff: circle, pypi, bumpr...

File: udata_croquemort/utils.py (hunk 7 above)
Comment (abulte) [reply udata_croquemort/utils.py]: Actually `JSONDecodeError` inherits from `ValueError` and this elegant piece of code would catch it when the response is (still) malformed.

File: udata_croquemort/tests.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+
+CROQUEMORT_TEST_URL = 'http://check.test'
+CHECK_ONE_URL = '{0}/check/one'.format(CROQUEMORT_TEST_URL)
+METADATA_URL = '{0}/url'.format(CROQUEMORT_TEST_URL)
+
+
+def metadata_factory(url, data=None):
+    """Base for a mocked Croquemort HTTP response"""
+    response = {
+      'etag': '',
```
Comment (noirbizarre) [inline udata_croquemort/tests.py]: 4 spaces ident

File: udata_croquemort/utils.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+import logging
+import time
+
+import requests
+
+from flask import current_app
+
+log = logging.getLogger(__name__)
+
+# See http://docs.python-requests.org/en/latest/user/advanced/#timeouts
```
Comment (noirbizarre) [inline udata_croquemort/utils.py]: :+1: :ok_hand:

File: udata_croquemort/utils.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+        try:
+            response = requests.get(retrieve_url, timeout=TIMEOUT)
+        except requests.Timeout:
+            raise UnreachableLinkChecker(TIMEOUT_LOG_MSG)
+        except requests.RequestException:
+            log.error(ERROR_LOG_MSG, exc_info=True)
+            raise UnreachableLinkChecker('{}: {}'.format(ERROR_LOG_MSG, e))
+        time.sleep(delay)
+        attempts += 1
+    return response.json()
```
Comment (noirbizarre) [inline udata_croquemort/utils.py]: May be protect from JSON parse error.

Comment (abulte) [reply udata_croquemort/utils.py]: Done by `is_pending`

**Comment by abulte** [comment]
@davidbgk OK for you?

**Pull request merged**